### PR TITLE
avoid allocating collection for intermediate certificates

### PR DESCRIPTION
**Pull request by wfurt**
There are actually two collection we will save in this change for each TLS handshake:

Current code always allocates `X509Certificate2Collection` to store intermediate certificates from handshake just to add all of them to X509Chain before validation. This change fixes the internal API so that we allocate X509Chain as needed and we add any intermediate certificates to it directly without need to allocate extra collection. (using ref X509Chain?)

The second collocation comes surprisingly from `null` check:
```c#
if (_chainPolicy != null && _chainPolicy.CustomTrustStore != null)
```

does not really work well with 
```c#
 public X509Certificate2Collection CustomTrustStore => _customTrustStore ??= new X509Certificate2Collection();
```

`CustomTrustStore` will never be `null` and the check will allocate in typical case.

### CODE REVIEW

File: src/libraries/Common/src/Interop/Windows/SChannel/UnmanagedCertificateContext.IntPtr.cs [hunk 1]
```diff
@@ -47,7 +45,7 @@ internal static unsafe X509Certificate2Collection GetRemoteCertificatesFromStore
                 }
             }
 
-            return result;
+            return;
```
Comment (rzikm) [inline src/libraries/Common/src/Interop/Windows/SChannel/UnmanagedCertificateContext.IntPtr.cs:48]: This return is unnecessary.

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.cs [hunk 2]
```diff
@@ -2,9 +2,12 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Diagnostics;
+using System.Net.Security;
 using System.Security;
 using System.Security.Cryptography;
 using System.Security.Cryptography.X509Certificates;
+using System.Runtime.InteropServices;
+
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.cs:10]: extra empty line?

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.cs [hunk 3]
```diff
@@ -14,6 +17,14 @@ internal static partial class CertificateValidationPal
...
         private static volatile X509Store? s_myCertStoreEx;
         private static volatile X509Store? s_myMachineCertStoreEx;
+        private static X509Chain? s_chain;
+
+        internal static X509Certificate2? GetRemoteCertificate(SafeDeleteContext securityContext) =>
+            GetRemoteCertificate(securityContext, retrieveChainCertificates: false, ref s_chain);
+
+        internal static X509Certificate2? GetRemoteCertificate(SafeDeleteContext securityContext, ref X509Chain? chain) =>
+            GetRemoteCertificate(securityContext, retrieveChainCertificates: true, ref chain);
+
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.cs:27]: extra empty line?

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Android.cs [hunk 4]
```diff
@@ -84,6 +64,11 @@ internal static partial class CertificateValidationPal
             }
             else
             {
+                if (chain == null)
+                {
+                    chain = new X509Chain();
+                }
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Android.cs:70]:
```suggestion
                chain ??= new X509Chain();
```

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.OSX.cs [hunk 5]
```diff
@@ -91,12 +70,17 @@ internal static partial class CertificateValidationPal
             {
                 long chainSize = Interop.AppleCrypto.X509ChainGetChainSize(chainHandle);
 
-                if (remoteCertificateStore != null)
+                if (retrieveChainCertificates)
                 {
+                    if (chain == null)
+                    {
+                        chain = new X509Chain();
+                    }
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.OSX.cs:78]:
```suggestion
                    chain ??= new X509Chain();
```

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs [hunk 6]
```diff
@@ -64,8 +45,13 @@ internal static partial class CertificateValidationPal
                     result = new X509Certificate2(remoteContext.DangerousGetHandle());
                 }
 
-                if (remoteCertificateStore != null)
+                if (retrieveChainCertificates)
                 {
+                    if (chain == null)
+                    {
+                        chain = new X509Chain();
+                    }
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs:53]:
```suggestion
                    chain ??= new X509Chain();
```

File: src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs [hunk 7]
```diff
@@ -72,9 +64,14 @@ internal static partial class CertificateValidationPal
...
                 if (remoteContext != null && !remoteContext.IsInvalid)
                 {
-                    if (retrieveCollection)
+                    if (retrieveChainCertificates)
                     {
-                        remoteCertificateCollection = UnmanagedCertificateContext.GetRemoteCertificatesFromStoreContext(remoteContext);
+                        if (chain == null)
+                        {
+                            chain = new X509Chain();
+                        }
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/CertificateValidationPal.Windows.cs:72]:
```suggestion
                        chain ??= new X509Chain();
```

File: src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs [hunk 8]
```diff
@@ -1016,18 +1014,17 @@ internal bool VerifyRemoteCertificate(RemoteCertificateValidationCallback? remot
                 }
                 else
                 {
-                    chain = new X509Chain();
+                    if (chain == null)
+                    {
+                        chain = new X509Chain();
+                    }
```
Comment (rzikm) [inline src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs:1020]:
```suggestion
                    chain ??= new X509Chain();
```

**Review by rzikm — APPROVED** [review]
LGTM, modulo comments.

File: src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs [hunk 9]
```diff
@@ -275,8 +275,8 @@ private static void OnRequestSendingRequest(WinHttpRequestState state)
                 }
 
                 // Get any additional certificates sent from the remote server during the TLS/SSL handshake.
-                X509Certificate2Collection remoteCertificateStore =
-                    UnmanagedCertificateContext.GetRemoteCertificatesFromStoreContext(certHandle);
+                X509Certificate2Collection remoteCertificateStore = new X509Certificate2Collection();
+                    UnmanagedCertificateContext.GetRemoteCertificatesFromStoreContext(certHandle, remoteCertificateStore);
```
Comment (campersau) [inline src/libraries/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs:279]:
```suggestion
                UnmanagedCertificateContext.GetRemoteCertificatesFromStoreContext(certHandle, remoteCertificateStore);
```

**Pull request merged**